azure-proto-core/Resources/UserAssignedIdentity.cs

```diff
@@ -0,0 +1,118 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+namespace azure_proto_core.Resources
+{
+    using System;
+    using System.Text.Json;
+    using Azure.Core;
+
+    public class UserAssignedIdentity
+    {
+        public Guid? ClientId { get; set; }
+
+        public Guid? PrincipalId { get; set; }
+
+        public UserAssignedIdentity() { }
+
+        public UserAssignedIdentity(Guid clientId, Guid principalId)
+        {
+            ClientId = clientId;
+            PrincipalId = principalId;
+        }
+
+        public int CompareTo(UserAssignedIdentity other)
+        {
+            if (other == null)
+                return 1;
+
+            int compareResult = 0;
+            if ((compareResult = PrincipalId.GetValueOrDefault().CompareTo(other.PrincipalId.GetValueOrDefault())) == 0 &&
+                (compareResult = ClientId.GetValueOrDefault().CompareTo(other.ClientId.GetValueOrDefault())) == 0)
+            {
+                return 0;
+            }
+
+            return compareResult;
+        }
+
+        public bool Equals(UserAssignedIdentity other)
+        {
+            if (other == null)
+                return false;
+
+            return ClientId.Equals(other.ClientId) && PrincipalId.Equals(other.PrincipalId);
+        }
+
+        public static UserAssignedIdentity Deserialize(JsonElement element)
+        {
+            if (element.ValueKind == JsonValueKind.Undefined)
```

Comment:
I may have missed this in system identity, but I believe we should have 2 exceptions here
```suggestion
            if (element == null)
                throw new ArgumentNullException(nameof(element));
                
            if (element.ValueKind == JsonValueKind.Undefined)
                throw new ArgumentException("JsonElement cannot be undefined", nameof(element)); 
```